kyverno/kyverno
# feat: skip generating VAP when an exception is defined

## Explanation
This PR skips generating ValidatingAdmissionPolicy in case an exception is created



## Related issue
Closes #8733


## Milestone of this PR
/milestone 1.12


## Documentation (optional)

My PR contains new or altered behavior to Kyverno. 
  
  

## What type of PR is this
/kind feature


## Proposed Changes



### Proof Manifests
1. Create a Kyverno policy:
```
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: disallow-host-path
spec:
  validationFailureAction: Audit
  background: false
  rules:
    - name: host-path
      match:
        all:
        - resources:
            kinds:
            - Deployment
            - StatefulSet
            operations:
            - CREATE
            - UPDATE
            selector:
              matchLabels:
                app: critical
      validate:
        cel:
          expressions:
            - expression: "!has(object.spec.template.spec.volumes) || object.spec.template.spec.volumes.all(volume, !has(volume.hostPath))"
              message: "HostPath volumes are forbidden. The field spec.template.spec.volumes[*].hostPath must be unset."
```
2. Check the generated VAP and its binding:
```
$ kubectl get validatingadmissionpolicybindings.admissionregistration.k8s.io 
NAME                         POLICYNAME           PARAMREF   AGE
disallow-host-path-binding   disallow-host-path   <unset>    18s

$ kubectl get validatingadmissionpolicy
NAME                 VALIDATIONS   PARAMKIND   AGE
disallow-host-path   1             <unset>     22s
```
3. check the Kyverno policy status:
```
status:
  validatingadmissionpolicy:
    generated: true
    message: ""
```
4. Create a policy exception:
```
apiVersion: kyverno.io/v2beta1
kind: PolicyException
metadata:
  name: delta-exception
  namespace: delta
spec:
  exceptions:
  - policyName: disallow-host-path
    ruleNames:
    - host-path
  match:
    any:
    - resources:
        kinds:
        - Deployment
        namespaces:
        - delta
        names:
        - important-tool*
```
5. check if there is a corresponding VAP/binding:
```
$ kubectl get validatingadmissionpolicy
No resources found

$ kubectl get validatingadmissionpolicybindings
No resources found
```
6. check the Kyverno policy status:
```
status:
  validatingadmissionpolicy:
    generated: false
    message: 'skip generating ValidatingAdmissionPolicy: a policy exception is configured.'
```


## Checklist




## Further Comments
